src/walk-manifest.js

```diff
@@ -49,116 +58,192 @@ var parseManifest = function(content) {
   return parser.manifest;
 };
 
-var parseKey = function(basedir, decrypt, resources, manifest, parent) {
-	if (!manifest.parsed.segments[0] || !manifest.parsed.segments[0].key) {
-		return {};
-	}
-	var key = manifest.parsed.segments[0].key;
-
-	var keyUri = key.uri;
-	if (!isAbsolute(keyUri)) {
-		keyUri = joinURI(path.dirname(manifest.uri), keyUri);
-	}
-
-	// if we are not decrypting then we just download the key
-	if (!decrypt) {
-		// put keys in parent-dir/key-name.key
-		key.file = basedir;
-		if (parent) {
-			key.file = path.dirname(parent.file);
-		}
-		key.file = path.join(key.file, fsSanitize(path.basename(key.uri)));
-
-		manifest.content = new Buffer(manifest.content.toString().replace(
-			key.uri,
-			path.relative(path.dirname(manifest.file), key.file)
-		));
-		key.uri = keyUri;
-		resources.push(key);
-		return key;
-	}
-
-	// get the aes key
-	var keyContent = syncRequest('GET', keyUri).getBody();
-	key.bytes = new Uint32Array([
-		keyContent.readUInt32BE(0),
-		keyContent.readUInt32BE(4),
-		keyContent.readUInt32BE(8),
-		keyContent.readUInt32BE(12)
-	]);
-
-	// remove the key from the manifest
-	manifest.content = new Buffer(manifest.content.toString().replace(
-		new RegExp('.*' + key.uri + '.*'),
-		''
-	));
-
-
-	return key;
+var parseKey = function(basedir, decrypt, resources, manifest, parent, callback) {
+  if (!manifest.parsed.segments[0] || !manifest.parsed.segments[0].key) {
+    return callback(null, {});
+  }
+  var key = manifest.parsed.segments[0].key;
+
+  var keyUri = key.uri;
+  if (!isAbsolute(keyUri)) {
+    keyUri = joinURI(path.dirname(manifest.uri), keyUri);
+  }
+
+  // if we are not decrypting then we just download the key
+  if (!decrypt) {
+    // put keys in parent-dir/key-name.key
+    key.file = basedir;
+    if (parent) {
+      key.file = path.dirname(parent.file);
+    }
+    key.file = path.join(key.file, fsSanitize(path.basename(key.uri)));
+
+    manifest.content = new Buffer(manifest.content.toString().replace(
+      key.uri,
+      path.relative(path.dirname(manifest.file), key.file)
+    ));
+    key.uri = keyUri;
+    resources.push(key);
+    return callback(null, key);
+  }
+
+  // get the aes key
+  request({url: keyUri, encoding: null, timeout: 1500}, function(error, response, body) {
+    if (error) {
+      const keyError = new Error(error.message + '|' + keyUri);
+      console.error(keyError, error);
+      return callback(keyError, {});
```

Comment:
no need to pass a second argument when we have an error, calling code shouldn't assume that data was provided if an error was passed.